src/loaders/common/callback_server_wrapper.py

```diff
@@ -17,69 +17,89 @@ class Conf:
     ) -> None:
         """
         Initialize the configuration class.
 
         Args:
             job_dir (str): The directory for SDK jobs per user.
             output_dir (str): The directory for a specific workspace id under sourcedata/ws.
-            worker_function (Callable): The function that will be called by the workers.
             kb_base_url (str): The base url of the KBase services.
-            token_filepath (str): The file path that stores a KBase token appropriate for the KBase environment. 
+            token_filepath (str): The file path that stores a KBase token appropriate for the KBase environment.
                                 If not supplied, the token must be provided in the environment variable KB_AUTH_TOKEN.
+            au_service_ver (str): The service verison of AssemblyUtilClient.
             workers (int): The number of workers to use for multiprocessing.
+            max_task (int): The maxmium subtasks for the callback server.
+            worker_function (Callable): The function that will be called by the workers.
             retrieve_sample (bool): Whether to retrieve sample for each genome object.
             ignore_no_sample_error (bool): Whether to ignore the error when no sample data is found.
+            workspace_downloader (bool): Whether to be used for the workspace downloader script.
         """
+        # common instance variables
+        ipv4 = loader_helper.get_ip()
         port = loader_helper.find_free_port()
         token = loader_helper.get_token(token_filepath)
-        self.retrieve_sample = retrieve_sample
-        self.ignore_no_sample_error = ignore_no_sample_error
-        ipv4 = loader_helper.get_ip()
+
+        ws_url = os.path.join(kb_base_url, "ws")
+        callback_url = "http://" + ipv4 + ":" + str(port)
+        print("callback_url:", callback_url)
+
         self._start_callback_server(
             docker.from_env(),
             uuid.uuid4().hex,
             job_dir,
             kb_base_url,
             token,
             port,
+            max_task,
             ipv4,
         )
 
-        ws_url = os.path.join(kb_base_url, "ws")
-        sample_url = os.path.join(kb_base_url, "sampleservice")
-        callback_url = "http://" + ipv4 + ":" + str(port)
-        print("callback_url:", callback_url)
-
         self.ws = Workspace(ws_url, token=token)
-        self.asu = AssemblyUtil(callback_url, token=token)
-        self.ss = SampleService(sample_url, token=token)
+        self.asu = AssemblyUtil(callback_url, service_ver=au_service_ver, token=token)
 
-        self.workers = workers
         self.output_dir = output_dir
-        self.input_queue = Queue()
-        self.output_queue = Queue()
         self.job_data_dir = loader_helper.make_job_data_dir(job_dir)
-        self.pools = Pool(workers, worker_function, [self])
+
+        self.logging = None
+
+        # unique to downloader
+        if workspace_downloader:
```

Comment:
It's pretty clear to me that this class is doing too much and should just be responsible for starting up the callback server and presenting necessary info (like the url, token) etc. to clients of the class. Task specific code should be moved to the class clients. That seems like it could be a chunk of work though. @Tianhao-Gu what do you think?

Reply:
Maybe start by doing that for the uploader since we're working on that right now and it's pretty simple

Reply:
so we are now just updating workspace_uploader.py file?

Reply:
I'd be interested in @Tianhao-Gu 's opinion, but the uploader certainly looks like less work based on a cursory perusal

Reply:
Moving task specific code to client classes is a logical step. But the effort required to update the Worksapce client is a lot more work than updating AssemblyUtil. I think it makes sense for now that we have a workspace wrapper so that we will have something workable sooner. And, from a performance perspective, I don't think there will be many improvements by moving code to Workspace client.

Reply:
Moving task specific code to client classes is a logical step. But the effort required to update the Worksapce client is a lot more work than updating AssemblyUtil. I think it makes sense for now that we have a workspace wrapper so that we will have something workable sooner. And, from a performance perspective, I don't think there will be many improvements by moving code to Workspace client.

Reply:
oh I see. I am completely on board with that. That's also what I meant by 'workspace wrapper'.

Reply:
It's only executed when `workspace_downloader` is `True`, so I'd assume so unless I'm missing something

Reply:
If we move that line, then current downloader parallelization will break. And fixing this parallelization problem would be a separate PR itself, unless I am missing something trivial.

Reply:
That's why I was saying maybe it should just be dune for the uploader which naively seems a lot simpler

Reply:
If so we could just add an issue for doing the downloader

Reply:
1. Updated the uploader
2. Added an issue for the downloader
https://github.com/kbase/collections/issues/572

Reply:
:+1: